[Backtracing] Add _Backtracing library to the build.

al45tair opened the pull request:
Adds a new swift_Backtracing library, with a corresponding _Backtracing module, to the build.  Also add some tests.

This is not public API at this point, but will be used by the external backtracing program, `swift-backtrace`.

rdar://104336548

al45tair committed:
[Backtracing] Add _Backtracing library to the build.

Adds a new swift_Backtracing library, with a corresponding _Backtracing
module, to the build.  Also add some tests.

This is not public API at this point, but will be used by the external
backtracing program, `swift-backtrace`.

rdar://104336548

al45tair commented:
This is part of a set of stacked PRs:

|  | PR                | Title                                                                      | Merges Into       |
|--|:------------------|:---------------------------------------------------------------------------|:------------------|
|  | apple/swift#63650 | [Backtracing] Add support for building target executables into libexec     | main              |
|  | al45tair/swift#8  | [Backtracing] Add support for looking up paths for auxiliary executables   | apple/swift#63650 |
|  | al45tair/swift#9  | [Backtracing][Runtime][Darwin] Add code to trigger the external backtracer | al45tair/swift#8  |
|👉| al45tair/swift#10 | [Backtracing] Add _Backtracing library to the build                        | al45tair/swift#9  |
|  | al45tair/swift#11 | [Backtracing][Frontend] Add support for implicit import of _Backtracing    | al45tair/swift#10 |
|  | al45tair/swift#12 | [Backtracing] Add swift-backtrace to the build                             | al45tair/swift#11 |
|  | al45tair/swift#13 | [Backtracing][Docs] Add an explanation of the workings of the backtracer   | al45tair/swift#12 |
|  | al45tair/swift#14 | [Backtracing] Add control over symbol caching                              | al45tair/swift#13 |

(Only the bottom PR will be able to run PR checks at any time. Please don't wait for the checks before review.)

weissi commented on stdlib/public/Backtracing/Backtrace.swift:
```diff
@@ -0,0 +1,496 @@
+
+  /// Holds information about the shared cache.
+  public struct SharedCacheInfo: Sendable {
+    #if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
```
This is public API, right? Shouldn't we aim for API compat between the platforms and have these available elsewhere? Otherwise we force users into this ugly `#if os` stuff.

al45tair replied:
Amusingly that's exactly what I thought about the `struct` itself, but I'd decided the fields were things that were only applicable to Darwin. That is, I wanted to make sure that anyone writing code against some other platform made certain to handle the `SharedCacheInfo`, but I didn't want to expose the fields there because they aren't meaningful.

I could make them `internal` I suppose, but the downside of that is that there's a chance on Darwin that you might want to display their contents from something outside of this code.

al45tair replied:
Decided to make the fields accessible but never return an object of this type, which makes more things `Optional` elsewhere but will make the API the same everywhere.

weissi commented on stdlib/public/Backtracing/Backtrace.swift:
```diff
@@ -0,0 +1,496 @@
+      var cacheInfo = dyld_process_cache_info()
+      _dyld_process_info_get_cache(dyldInfo, &cacheInfo)
+      let theUUID = withUnsafePointer(to: cacheInfo.cacheUUID){
+        $0.withMemoryRebound(to: UInt8.self,
```
You don't need the rebinding here. You can

```
Array(UnsafeRawBufferPointer(start: $0, count: MemoryLayout<uuid_t>.size))
```

al45tair replied:
Makes sense that would work, and saves rebinding memory.

weissi commented on stdlib/public/Backtracing/BacktraceFormatter.swift:
```diff
@@ -0,0 +1,983 @@
+
+    while feof(fp) == 0 && ferror(fp) == 0 {
+      guard let result = fgets(buffer.baseAddress,
+                               Int32(buffer.count), fp) else {
```
I think `CInt` is more appropriate than `Int32` here (and other places too)

al45tair replied:
Yeah, I think this is a result of me leaning on the compiler's fixit messages to catch these. I'll do a pass through and see if I can pick them all up.

al45tair replied:
I fixed this one, but I wasn't sure where else you had in mind.

weissi replied:
Can't find it immediately either. Maybe I'm wrong and there was just this one and I got confused. Let's consider this done.

weissi commented on stdlib/public/Backtracing/CoreSymbolication.swift:
```diff
@@ -0,0 +1,360 @@
+// CFString.  As a result, we need to do the dance manually.
+
+private func toCFString(_ s: String) -> CFString! {
+  let bytes = Array<UInt8>(s.utf8)
```
you can save an alloc + copy here if you write

```
var s = s
s.withUTF8 { ... }
```

then String will just vend you its internals (already utf8).

weissi replied:
(unless bridged from NSString in which case we'll alloc & copy)

weissi commented on stdlib/public/Backtracing/CoreSymbolication.swift:
```diff
@@ -0,0 +1,360 @@
+
+  if let ptr = CFStringGetCStringPtr(cf,
+                                     CFStringBuiltInEncodings.ASCII.rawValue) {
+    return ptr.withMemoryRebound(to: UInt8.self, capacity: length) {
```
no need to rebind here, do

```
String(decoding: UnsafeRawBufferPointer(start: ptr, count: length), as: UTF8.self)
```

weissi commented on stdlib/public/Backtracing/MemoryReader.swift:
```diff
@@ -0,0 +1,118 @@
+  private var task: task_t
+
+  // Sadly we can't expose the type of this argument
+  public init(task: Any) {
```
Could we have an opaque `VMTask` type and then

```
extension task_t {
   var asTask: VMTask {
      return VMTask(...)
   }
}
```

al45tair replied:
Sadly, I don't think we can, the reason being that we can't add a public extension on `task_t` from `MemoryReader.swift` because `task_t` is from `Darwin.Mach`, which is imported `@implementationOnly`.

weissi reviewed (commented):
Looks great!

al45tair committed:
[Backtracing][Runtime] Fix Linux build to use getauxval(AT_SECURE).

Linux doesn't have `issetugid()`, so use `getauxval(AT_SECURE)` there
instead.

rdar://105391747

al45tair committed:
[Backtracing] Various improvements following Johannes' remarks.

Removed some unnecessary memory rebinding.

Made `CFString` conversion slightly more efficient.

Provide the `SharedCacheInfo` fields everywhere, but make it optional all
over as well.

rdar://104336548

mikeash commented on stdlib/public/Backtracing/Backtrace.swift:
```diff
@@ -0,0 +1,491 @@
+  ///
+  /// This will be `UInt32` or `UInt64` on current platforms.
+  #if arch(x86_64) || arch(arm64)
+  public typealias Address = UInt64
```
Can we unconditionally use `UInt` instead? That will always match the platform pointer size.

al45tair replied:
I think we probably can. There's some ongoing discussion in the forums about this, and I've flip-flopped a few times about whether I want this to be a fixed type so you can use the `Backtrace` struct to load backtraces from other platforms or whether I want it to be a platform specific type. For now I think I'll change it to `UInt` and keep pondering what to do for the official version of the API when we declare it done and enable implicit import for this.

mikeash commented on stdlib/public/Backtracing/Backtrace.swift:
```diff
@@ -0,0 +1,491 @@
+
+  /// Represents an individual frame in a backtrace.
+  public enum Frame: CustomStringConvertible, Sendable {
+    /// An accurate program counter.
```
"Accurate" confuses me a bit here, as I'm not sure what the alternative is. I'd like a little explanation of that, or just drop that word.

al45tair replied:
"Accurate" means that it points to the instruction responsible; the alternative is "return address", which instead points to the instruction *after* the subroutine call, which means that when symbolicating we need to adjust the value (otherwise we'll get the backtrace wrong in some cases). I'll add more explanation in the comment.

al45tair replied:
I've updated the comment. On reflection I decided to remove "accurate" and add more explanation below.

mikeash commented on stdlib/public/Backtracing/Backtrace.swift:
```diff
@@ -0,0 +1,491 @@
+  ///
+  /// Some backtracing algorithms may require this information, in which case
+  /// it will be filled in by the `capture()` method.  Other algorithms may
+  /// not, in which case it will be empty and you can capture an image list
```
Is it actually `nil` when the backtracing algorithm doesn't require the information?

al45tair replied:
Yes. I'll fix the comment.

mikeash commented on stdlib/public/Backtracing/Backtrace.swift:
```diff
@@ -0,0 +1,491 @@
+  }
+
+  @_spi(Internal)
+  public static func capture(from context: some Context,
```
Does this also want an `@inline` attribute to make it consistent whether it shows up as a separate frame?

al45tair replied:
No, because the `Context` has already been captured at this point. It's context capture that matters from that perspective.

mikeash commented on stdlib/public/Backtracing/Backtrace.swift:
```diff
@@ -0,0 +1,491 @@
+          .dropFirst(offset)
+
+        if let limit = limit {
+          if limit == 0 {
```
`<= 0` will let this fail gracefully rather than crashing strangely if the caller is weird and passes a negative number.

al45tair replied:
Good shout.

mikeash commented on stdlib/public/Backtracing/Backtrace.swift:
```diff
@@ -0,0 +1,491 @@
+    #if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
+    let task = process as! task_t
+
+    withDyldProcessInfo(for: task){ dyldInfo in
```
```suggestion
    withDyldProcessInfo(for: task) { dyldInfo in
```

mikeash commented on stdlib/public/Backtracing/Backtrace.swift:
```diff
@@ -0,0 +1,491 @@
+    return withDyldProcessInfo(for: task){ dyldInfo in
+      var cacheInfo = dyld_process_cache_info()
+      _dyld_process_info_get_cache(dyldInfo, &cacheInfo)
+      let theUUID = withUnsafePointer(to: cacheInfo.cacheUUID){
```
```suggestion
      let theUUID = withUnsafePointer(to: cacheInfo.cacheUUID) {
```

al45tair replied:
I picked up a couple more of these by searching for `){`.

mikeash commented on stdlib/public/Backtracing/BacktraceFormatter.swift:
```diff
@@ -0,0 +1,983 @@
+  /// @param enabled  If true, we show demangled names if we have them.
+  ///
+  /// @returns A new `BacktraceFormattingOptions` structure.
+  public static func demangle(_ enabled: Bool = true) -> BacktraceFormattingOptions {
```
WTB macros.

al45tair replied:
Yep. I kept pondering whether there was a better way to write this that I'd missed, but I don't think there is (without macros).

mikeash replied:
You could get close, but with a bit different syntax, by making a general-purpose generic function that takes a key path and a value. Not quite the same though.

mikeash commented on stdlib/public/Backtracing/BacktraceFormatter.swift:
```diff
@@ -0,0 +1,983 @@
+    return 1
+  }
+
+  if #available(macOS 10.12.2, *) {
```
Xcode no longer officially supports deployment targets below 10.13 so this can probably be dropped.

mikeash commented on stdlib/public/Backtracing/FramePointerUnwinder.swift:
```diff
@@ -0,0 +1,152 @@
+    #if (os(macOS) || os(iOS) || os(watchOS)) && (arch(arm64) || arch(arm64_32) || arch(x86_64))
+    // On Darwin, we borrow a bit of the frame pointer to indicate async
+    // stack frames
+    return (storedFp & (1 << 60)) != 0 && _getCurrentAsyncTask() != nil
```
This might be wrong on ARM64_32, where as I recall we sort of treat the frame pointer as a 64-bit quantity for the purposes of setting the async frame bit, but it's mostly 32-bit.

al45tair replied:
I checked against the code `libc` uses for the same thing; if it's wrong here, it's wrong there too (so at least we're consistent).

al45tair committed:
[Backtracing][Runtime] Fix a copy/paste blunder.

Accidentally wrote `thread_resume()` instead of `thread_suspend()`.

rdar://105391747

al45tair committed:
[Backtracing] Tweaks after Mike's remarks.

Just use `UInt` for `Address`.  This is still the subject of some discussion
on the forums, but I haven't decided precisely what to do about it and `UInt`
makes sense for now.  This also necessitated some casts elsewhere.

Improve some comments.

Made `limit` robust against silly negative values.

A couple of formatting fixes.

Don't bother supporting the macOS 10.12.2 SDK as Xcode now supports a minimum
of 10.13.

al45tair committed:
[Backtracing] Fix Windows assembler related problem.

We need to enable MASM/MARMASM and we need to add a get-cpu-context.asm
file for it to build.

rdar://104336548

al45tair committed:
[Backtracing][Windows] Use CRT, not MSVCRT.

The correct name for the runtime library appears to be CRT, not MSVCRT.

rdar://104336548

al45tair committed:
[Backtracing] Fix OS conditionals so iOS builds work.

Some of the `TARGET` and `os()` conditionals needed to be updated.

rdar://104336548

al45tair committed:
[Backtracing] Don't try to include <libproc.h> for iOS.

iOS doesn't have <libproc.h>.  We don't need it anyway there.

rdar://104336548

al45tair committed:
[Backtracing] Only include libproc.h if it's present.

rdar://104336548
